pkg/tfgen/internal/unrec/package_spec_util.go

```diff
@@ -0,0 +1,221 @@
+// Copyright 2016-2024, Pulumi Corporation.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package unrec
+
+import (
+	"encoding/json"
+	"fmt"
+	"strings"
+
+	pschema "github.com/pulumi/pulumi/pkg/v3/codegen/schema"
+	"github.com/pulumi/pulumi/sdk/v3/go/common/tokens"
+	"github.com/pulumi/pulumi/sdk/v3/go/common/util/contract"
+)
+
+func parseLocalRef(rawRef string) (tokens.Type, bool) {
+	if !strings.HasPrefix(rawRef, "#/types/") {
+		return "", false
+	}
+	cleanRef := strings.TrimPrefix(rawRef, "#/types/")
+	return tokens.Type(cleanRef), true
```

Comment:
Is there a reason not to validate here?
```suggestion
	typ, err := tokens.ParseTypeToken(cleanRef)
	contract.AssertNoErrorf(err, "invalid schema")
	return typ, true
```